### PR TITLE
Warn if postgres database has non-C locale.

### LINKED ISSUE #6722: Refuse to run against postgres with a non-C collation
A number of people have reported (#6696) database index corruption which basically stems from their databases having a UTF-8 collation setting.

Our [documentation](https://github.com/matrix-org/synapse/blob/release-v1.8.0/docs/postgres.md) tells you to set the collation to "C" to avoid this sort of thing (and possibly improve performance, according to [postgres-general](https://www.postgresql.org/message-id/f2f4c44f-5673-496b-b42b-cf63ba483a5e%40manitou-mail.org)), but this is little comfort to people who missed that advice and now have corrupt indexes to deal with.

By way of a preventative measure, maybe we should complain if somebody is using a different collation?

```
synapse=# SELECT datcollate FROM pg_database WHERE datname = current_database();
 datcollate 
------------
 C
(1 row)
```

### PR DESCRIPTION
**Pull request by erikjohnston**
As using non-C locale can cause issues on upgrading OS.

Fixes #6722.

---

I think we probably want to have better wording here, suggestions welcome.

### CODE REVIEW

**Comment by aaronraimist** [comment]
Might be good to add a note to `UPGRADE.md` explain [what you need to do](https://github.com/matrix-org/synapse/issues/6696#issuecomment-574706305) if you get this warning.

File: changelog.d/6734.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Warn if postgrse database has a non-C locale, as that can cause issues when upgrading locales (e.g. due to upgrading OS).
```
Comment (turt2live) [inline changelog.d/6734.bugfix]:
```suggestion
Warn if postgres database has a non-C locale, as that can cause issues when upgrading locales (e.g. due to upgrading OS).
```

File: docs/postgres.md [hunk 2]
```diff
@@ -63,6 +63,20 @@ You may need to enable password authentication so `synapse_user` can
...
+`COLLATE` and `CTYPE` set, and will log warnings on existing databases. Using
+different locales can cause issues if the locale library is updated from
+underneath the database, or if a different version of the locale is used on any
+replicas.
+
+The safest way to fix the issue is to take a dump and recreate the database with
+the correct `COLLATE` and `CTYPE` parameters (as per
+[docs/postgres.md](docs/postgres.md)). It is also possible to change the
+parameters on a live database and run a `REINDEX` on the entire database,
+however extreme care must be taken to avoid database corruption.
```
Comment (richvdh) [inline docs/postgres.md]: won't either of these methods fail until you take steps to remove duplicate rows?

Comment (erikjohnston) [reply docs/postgres.md]: Oh, true. TBH I'm not really sure what to suggest they do other than a bland "delete them maybe?"

Comment (richvdh) [reply docs/postgres.md]: +1 to "delete them maybe"

btw the secret `ctid` pseudo-column can be useful for distinguishing similar rows

**Pull request merged**